crates/miden-tx/src/host/mod.rs

```diff
@@ -1129,6 +1213,61 @@ impl<'store, STORE> TransactionBaseHost<'store, STORE> {
     }
 }
 
+/// Extracts and validates note inputs from the advice provider using trial unhashing.
+///
+/// This function tries to determine the correct number of inputs by:
+/// 1. Finding the last non-zero element as a starting point
+/// 2. Building NoteInputs and checking if the hash matches inputs_commitment
+/// 3. If not, incrementing num_inputs and trying again (up to 6 more times)
+/// 4. If num_inputs grows to the size of inputs_data and there's still no match, returning an error
+fn extract_note_inputs(
+    process: &ProcessState,
+    inputs_commitment: &Word,
+) -> Result<NoteInputs, TransactionKernelError> {
+    let inputs_data = process.advice_provider().get_mapped_values(inputs_commitment);
+
+    let inputs = match inputs_data {
+        None => NoteInputs::default(),
+        Some(inputs) => {
+            // Start with the last non-zero element as a hint
+            let initial_num_inputs =
+                inputs.iter().rposition(|&x| x != ZERO).map(|pos| pos + 1).unwrap_or(0);
+
+            // Try different input counts using trial unhashing
+            let max_attempts = 7; // Initial attempt + 6 more
+            let mut num_inputs = initial_num_inputs;
+
+            for attempt in 0..max_attempts {
+                if num_inputs > inputs.len() {
+                    break;
+                }
```

Comment:
I think we can avoid having `max_attempts` variable here by doing something like:
```rust
while inputs.len() <= num_inputs {
    ...
}
```